Add CBMC proof for derive_data_key

adpaco-aws opened the pull request:
This PR adds a proof for derive_data_key, a function from session_decrypt.c, and also patches #653 for the proof to work.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

adpaco-aws commented:
@feliperodri @alex-chew I have added some content to this proof so that it conforms to the style followed by previous proofs (i.e., use "ensure" functions and similar). However, in this proof we do not need a valid `aws_cryptosdk_dec_materials`, only the `unencrypted_data_key` buffer contained within has to be valid. A similar thing happens with `aws_cryptosdk_session` (we only need to have assumptions for some header buffers or alg_properties.

Finally, not sure which postconditions would be godd to check for. In case of failure, `aws_raise_errors` shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.

alex-chew commented:
Mostly LGTM. I noticed some commented-out lines referencing `ensure_nondet_sig_ctx_has_allocated_members`. Does https://github.com/aws/aws-encryption-sdk-c/issues/667#issuecomment-744898996 make those lines work, or do we not need them (in which case we can simply remove them)?

> Finally, not sure which postconditions would be godd to check for. In case of failure, aws_raise_errors shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.

Can we not retrieve the raised error code using `aws_last_error`?

adpaco-aws commented:
> Mostly LGTM. I noticed some commented-out lines referencing `ensure_nondet_sig_ctx_has_allocated_members`. Does [#667 (comment)](https://github.com/aws/aws-encryption-sdk-c/issues/667#issuecomment-744898996) make those lines work, or do we not need them (in which case we can simply remove them)?

We did not need them but it is good to have all members allocated. Anyway, the issue is now fixed.

> > Finally, not sure which postconditions would be godd to check for. In case of failure, aws_raise_errors shadows the reason why it failed, so we cannot follow the assertion style used in previous "derive_key" functions.
> 
> Can we not retrieve the raised error code using `aws_last_error`?

Thanks for the tip! I am still not familiar with all these functions 😞 

I have added a couple post-conditions using it. More complex ones (like the ones on `aws_cryptosdk_private_derive_key`) can come from different functions and force us to reproduce part of the code in `derive_data_key` or other functions, so I do not think it is a good idea adding them.

alex-chew commented:
LGTM!

> I have added a couple post-conditions using it. More complex ones (like the ones on aws_cryptosdk_private_derive_key) can come from different functions and force us to reproduce part of the code in derive_data_key or other functions, so I do not think it is a good idea adding them.

That's fair, and the ones you added look good to me.

feliperodri commented on verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c:
```diff
@@ -0,0 +1,95 @@
+    struct aws_cryptosdk_dec_materials *materials = ensure_dec_materials_attempt_allocation();
+
+    /* Assumptions */
+    __CPROVER_assume(session != NULL);
+    __CPROVER_assume(aws_cryptosdk_commitment_policy_is_valid(session->commitment_policy));
+
+    __CPROVER_assume(session->alg_props != NULL);
+    __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(session->alg_props));
+    __CPROVER_assume(session->alg_props->commitment_len <= sizeof(session->key_commitment_arr));
+
+    __CPROVER_assume(aws_byte_buf_is_bounded(&session->header.message_id, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.message_id);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.message_id));
+
+    __CPROVER_assume(aws_byte_buf_is_bounded(&session->header.alg_suite_data, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.alg_suite_data);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.alg_suite_data));
```
Replace that by
```c
__CPROVER_assume(aws_cryptosdk_session_is_valid(session));
```

adpaco-aws replied:
We do have a `session_setup` function which boths performs allocation and common assumptions.

feliperodri commented on verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c:
```diff
@@ -0,0 +1,95 @@
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.alg_suite_data);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.alg_suite_data));
+
+    __CPROVER_assume(materials != NULL);
+    __CPROVER_assume(aws_byte_buf_is_bounded(&materials->unencrypted_data_key, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&materials->unencrypted_data_key);
+    __CPROVER_assume(aws_byte_buf_is_valid(&materials->unencrypted_data_key));
```
Replace that by
```c
__CPROVER_assume(aws_cryptosdk_dec_materials_is_valid(materials));
```

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
+struct aws_cryptosdk_session *ensure_session_attempt_allocation(size_t max_len) {
+    struct aws_cryptosdk_session *session = malloc(sizeof(struct aws_cryptosdk_session));
+    if (session) {
+        session->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+        session->cmm         = malloc(sizeof(struct aws_cryptosdk_cmm));
+        session->header_copy = malloc(sizeof(uint8_t));
+        session->alg_props   = ensure_alg_properties_attempt_allocation(max_len);
+        session->signctx     = ensure_nondet_sig_ctx_has_allocated_members();
```
```c
    struct aws_allocator *alloc;
    struct aws_cryptosdk_cmm *cmm;
    uint8_t *header_copy;
    const struct aws_cryptosdk_alg_properties *alg_props;
    struct aws_cryptosdk_sig_ctx *signctx;
```
Makes sure all are properly allocated and, if they also have pointers, their pointers are properly allocated.

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
+    struct aws_cryptosdk_session *session = malloc(sizeof(struct aws_cryptosdk_session));
+    if (session) {
+        session->alloc       = nondet_bool() ? NULL : can_fail_allocator();
+        session->cmm         = malloc(sizeof(struct aws_cryptosdk_cmm));
+        session->header_copy = malloc(sizeof(uint8_t));
```
```c
        session->cmm         = malloc(sizeof(*(session->cmm)));
        session->header_copy = malloc(sizeof(*(session->header_copy)));
```

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -253,3 +254,24 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
+    return session;
+}
+
+struct aws_cryptosdk_dec_materials *ensure_dec_materials_attempt_allocation() {
+    struct aws_cryptosdk_dec_materials *materials = malloc(sizeof(struct aws_cryptosdk_dec_materials));
+    if (materials) {
+        materials->alloc   = nondet_bool() ? NULL : can_fail_allocator();
+        materials->signctx = ensure_nondet_sig_ctx_has_allocated_members();
+    }
+    return materials;
+}
```
Same here.

feliperodri commented on verification/cbmc/sources/openssl/ec_override.c:
```diff
@@ -488,5 +488,11 @@ void write_unconstrained_data(unsigned char *out, size_t len) {
     // Currently we ignore the len parameter and just fill the entire buffer with unconstrained data.
     // This is fine because it is strictly more general behavior than writing only len bytes.
-    __CPROVER_havoc_object(out);
+    // __CPROVER_havoc_object(out);
+    if (len > 0) {
+        size_t index;
+        unsigned char nondet_char;
+        __CPROVER_assume(index < len);
+        out[index] = nondet_char;
+    }
```
Do we have a GitHub issue to track this issue? Both here and in our openssl verification model?

adpaco-aws replied:
Yes, we do: #653 https://github.com/awslabs/aws-verification-model-for-libcrypto/issues/18

tegansb reviewed (approved):
Overall, LGTM modulo the comments Felipe made about allocation. I'm going to go ahead and approve it assuming those get fixed because I don't want to prevent it getting merged because of time zones.

tegansb commented:
> A similar thing happens with `aws_cryptosdk_session` (we only need to have assumptions for some header buffers or alg_properties). Moreover, session is the largest struct in these proofs and I do not think we are going to a use its validating function on other proofs.
> 
Very similar to the sign_header and priv_try_gen_key proofs. Only some of the fields need to be allocated and valid for the proofs to work. I also currently check the validity of the fields independently but will use any helper functions you introduce here if they make sense for those proofs.

adpaco-aws commented on verification/cbmc/proofs/derive_data_key/derive_data_key_harness.c:
```diff
@@ -0,0 +1,95 @@
+
+    __CPROVER_assume(aws_byte_buf_is_bounded(&session->header.alg_suite_data, MAX_BUFFER_SIZE));
+    ensure_byte_buf_has_allocated_buffer_member(&session->header.alg_suite_data);
+    __CPROVER_assume(aws_byte_buf_is_valid(&session->header.alg_suite_data));
```
Thanks, I have included the functions used in `hdr_setup` in the new `session_setup` method.

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -79,13 +80,9 @@ void ensure_record_has_allocated_members(struct aws_cryptosdk_keyring_trace_reco
-            ensure_record_has_allocated_members(record, max_len);
-        }
+    for (size_t i = 0; i < trace->length; ++i) {
+        struct aws_cryptosdk_keyring_trace_record *data = (struct aws_cryptosdk_keyring_trace_record *)trace->data;
```
Could you move the declaration outside of the loop?

adpaco-aws replied:
Done!

feliperodri reviewed (approved):
LGTM modulo one comment.
